File: bytopic/linkedlist/java/group2/RemoveDuplicatesFromSortedListII.java
// Given the head of a sorted linked list, delete all nodes that have duplicate numbers, leaving only
// distinct numbers from the original list. Return the linked list sorted as well.
// Example :
// Input: head = [1,2,3,3,4,4,5]
// Output: [1,2,5]
//
// Tag: 64/150
// Tag: 82/2927, R918/2936 (overall frequency ranking)

class ListNode {
    int value;
    ListNode next;

    public ListNode(int value) {
        this.value = value;
    }
    public ListNode(int value, ListNode next) {
        this.value = value;
        this.next = next;
    }
}

class RemovalOfDuplicatesFromSortedListII {
    public static void main(String[] args) {
        ListNode list = new ListNode(1);
        list.next = new ListNode(2);
        list.next.next = new ListNode(3);
        list.next.next.next = new ListNode(3);
        list.next.next.next.next = new ListNode(4);
        list.next.next.next.next.next = new ListNode(4);
        list.next.next.next.next.next.next = new ListNode(5);

        ListNode result = deleteDuplicates(list);
        while (result.next != null) {
            System.out.print(result.value + ",");
            result = result.next;
        }
        System.out.print(result.value); // Last value
        System.out.println();
    }

    static ListNode deleteDuplicates(ListNode head) {
        if (head == null || head.next == null)
            return head;

        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode curr = dummy;

        while (curr.next != null && curr.next.next != null) {
            if (curr.next.value == curr.next.next.value) {
                int duplicate = curr.next.value;
                while (curr.next != null && curr.next.value == duplicate) {
                    curr.next = curr.next.next;
                }
            } else {
                curr = curr.next;
            }
        }

        return dummy.next;
    }
}

File: bytopic/linkedlist/python/group2/remove_duplicates_from_sorted_list_ii.py
"""
Given the head of a sorted linked list, delete all nodes that have duplicate numbers, leaving only
distinct numbers from the original list. Return the linked list sorted as well.

Example :
Input: head = [1,2,3,3,4,4,5]
Output: [1,2,5]

Tag: 64/150
Tag: 82/2927, R918/2936 (overall frequency ranking)
"""
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def delete_duplicates(head: Optional[ListNode]) -> Optional[ListNode]:
    dummy = ListNode(0)
    dummy.next = head
    prev = dummy
    while head and head.next:
        if head.val == head.next.val:
            while head and head.next and head.val == head.next.val:
                head = head.next
            head = head.next
            prev.next = head
        else:
            prev = prev.next
            head = head.next
    return dummy.next


def main():
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(3)
    head.next.next.next.next = ListNode(4)
    head.next.next.next.next.next = ListNode(4)
    head.next.next.next.next.next.next = ListNode(5)

    res = delete_duplicates(head)
    print(res.val, res.next.val, res.next.next.val)


if __name__ == "__main__":
    main()

